### PR TITLE
feat: implement proxy handler

### PR DESCRIPTION
**Pull request by abc3**
The current PR introduces a new handler that will replace session mode. It authenticates incoming connections via auth queries, establishes a TCP connection to the tenant's database, and then simply forwards data between the client and the database, functioning primarily as a proxy

- [x] cancel query
- [x] terminate auth_query connection
- [x] active once
- [x] application_name with "via supavisor"
- [x] handle client disconnect
- [x] handle DB disconnect
- [x] auth_query conn is terminated if the client is killed
- [x] telemetry
- [x] delete metrics  
- [x] tests

### CODE REVIEW

**Comment by abc3** [comment]
Should we consider using `defdelegate` in `Supavisor.Handlers.Proxy.Handler`?

File: lib/supavisor/handlers/proxy/client.ex [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+        ssl_opts: ssl_opts || []
+      )
+
+    # kill the postgrex connection if the current process exits unexpectedly
+    Process.link(conn)
+
+    msg =
+      "ProxyClient: Connected to db #{tenant.db_host} #{tenant.db_port} #{tenant.db_database} #{user.db_user}"
+
+    Logger.debug(msg)
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/client.ex:568]:
```suggestion
    Logger.debug("ProxyClient: Connected to db #{tenant.db_host} #{tenant.db_port} #{tenant.db_database} #{user.db_user}")
```

File: lib/supavisor/handlers/proxy/client.ex [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+
+    resp =
+      case H.get_user_secret(conn, tenant.auth_query, db_user) do
+        {:ok, secret} ->
+          t = if secret.digest == :md5, do: :auth_query_md5, else: :auth_query
+          {:ok, {t, fn -> Map.put(secret, :alias, user.db_user_alias) end}}
+
+        {:error, reason} ->
+          {:error, reason}
+      end
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/client.ex:578]:
```suggestion
      with {:ok, secret} <- H.get_user_secret(conn, tenant.auth_query, db_user) do
        t = if secret.digest == :md5, do: :auth_query_md5, else: :auth_query
        {:ok, {t, fn -> Map.put(secret, :alias, user.db_user_alias) end}}
      end
```

File: lib/supavisor/handlers/proxy/client.ex [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+
+    idle ++ heartbeat
+  end
+
+  @spec app_name(any()) :: String.t()
+  def app_name(name) when is_binary(name) do
+    suffix = " via Supavisor"
+    # https://www.postgresql.org/docs/current/runtime-config-logging.html#GUC-APPLICATION-NAME
+    max_len = 64
+    suffix_len = String.length(suffix)
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/client.ex:659]: Use `byte_size/1` there, or ideally we can use precomputed value, as this value is constant.

File: lib/supavisor/handlers/proxy/db.ex [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        ps_encoded = Server.encode_parameter_status(ps)
+
+        {:next_state, :idle, %{data | parameter_status: ps},
+         {:next_event, :internal, {:client, {:greetings, ps_encoded}}}}
+
+      other ->
+        Logger.error("ProxyDb: Undefined auth response #{inspect(other)}")
+        {:stop, :auth_error, data}
+    end
+  end
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/db.ex:156]: That is way to long and way to complex function to be a single one.

File: lib/supavisor/handlers/proxy/db.ex [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    HH.sock_close(data.sock)
+    {:stop, :db_socket_closed, data}
+  end
+
+  @spec try_ssl_handshake(S.tcp_sock(), map) :: {:ok, S.db_sock()} | {:error, term()}
+  def try_ssl_handshake(sock, %{upstream_ssl: true} = auth) do
+    case HH.sock_send(sock, Server.ssl_request()) do
+      :ok -> ssl_recv(sock, auth)
+      error -> error
+    end
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/db.ex:177]:
```suggestion
    with :ok <- HH.sock_send(sock, Server.ssl_request()) do
      ssl_recv(sock, auth)
    end
```

File: lib/supavisor/handlers/proxy/handler.ex [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+  def init(ref, trans, opts) do
+    Process.flag(:trap_exit, true)
+    Helpers.set_max_heap_size(90)
+
+    {:ok, sock} = :ranch.handshake(ref)
+    :ok = trans.setopts(sock, active: true)
+    Logger.debug("ClientHandler is: #{inspect(self())}")
+
+    data = %{
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/handler.ex:41]: I still think that `defstruct` for state would be useful.

File: lib/supavisor/handlers/proxy/handler.ex [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  end
+
+  def handle_event(event, {closed, sock} = msg, state, %{db_sock: {_, sock}} = data)
+      when closed in @sock_closed do
+    Db.handle_event(event, msg, state, data)
+  end
+
+  def terminate({:shutdown, reason}, state, data) do
+    msg = "ProxyHandler: Terminating with reason: #{inspect(reason)} when state was #{state}"
+    Logger.info(msg)
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/handler.ex:123]:
```suggestion
    Logger.info("ProxyHandler: Terminating with reason: #{inspect(reason)} when state was #{state}")
```

File: lib/supavisor/handlers/proxy/handler.ex [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    HH.sock_close(data.sock)
+    HH.sock_close(data.db_sock)
+    :ok
+  end
+
+  def terminate(reason, state, data) do
+    HH.sock_close(data.sock)
+    HH.sock_close(data.db_sock)
+    msg = "ProxyHandler: Terminating with reason: #{inspect(reason)} when state was #{state}"
+    Logger.info(msg)
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/handler.ex:134]:
```suggestion
    Logger.info("ProxyHandler: Terminating with reason: #{inspect(reason)} when state was #{state}")
```

File: lib/supavisor/helpers.ex [hunk 9]
```diff
@@ -357,4 +357,12 @@ defmodule Supavisor.Helpers do
     Logger.notice("Setting log level to #{inspect(level)}")
     Logger.put_process_level(self(), level)
   end
+
+  @spec peer_ip(:gen_tcp.socket()) :: String.t()
+  def peer_ip(socket) do
+    case :inet.peername(socket) do
+      {:ok, {ip, _port}} -> "#{:inet.ntoa(ip)}"
```
Comment (hauleth) [inline lib/supavisor/helpers.ex:364]:
```suggestion
      {:ok, {ip, _port}} -> List.to_string(:inet.ntoa(ip))
```

File: lib/supavisor/handlers/proxy/client.ex [hunk 10]
```diff
@@ -0,0 +1,673 @@
...
+  alias Supavisor.Helpers, as: H
+  alias Supavisor.HandlerHelpers, as: HH
+
+  alias Supavisor.{
+    Tenants,
+    ProxyHandlerDb,
+    Monitoring.Telem,
+    Protocol.Client,
+    Protocol.Server
+  }
```
[the remark this answers is not in the document]
Comment (abc3) [reply lib/supavisor/handlers/proxy/client.ex:17]: > I prefer having each alias independently

why? be able to find all imports? it's already possible with the language server

File: lib/supavisor/handlers/proxy/client.ex [hunk 11]
```diff
@@ -0,0 +1,673 @@
...
+
+  # # ssl request from client
+  def handle_event(:info, {:tcp, _, <<_::64>>}, :exchange, %{sock: sock} = data) do
+    Logger.debug("ProxyClient: Client is trying to connect with SSL")
+
+    downstream_cert = H.downstream_cert()
+    downstream_key = H.downstream_key()
+
+    # SSL negotiation, S/N/Error
+    if !!downstream_cert and !!downstream_key do
```
[the remark this answers is not in the document]
Comment (abc3) [reply lib/supavisor/handlers/proxy/client.ex:64]: 🤣

[the remark this answers is not in the document]
Comment (abc3) [reply lib/supavisor/handlers/proxy/client.ex:64]: IMO, simple boolean conversion

File: lib/supavisor/handlers/proxy/client.ex [hunk 12]
```diff
@@ -0,0 +1,673 @@
...
+
+    case db_pid_meta(key) do
+      [{^db_pid, meta}] ->
+        :ok = HH.cancel_query(meta.host, meta.port, meta.ip_ver, meta.pid, meta.key)
+
+      error ->
+        msg =
+          "ProxyClient: Received cancel but no proc was found #{inspect(key)} #{inspect(error)}"
+
+        Logger.error(msg)
```
[the remark this answers is not in the document]
Comment (abc3) [reply lib/supavisor/handlers/proxy/client.ex:50]: Yeah, this garbage is for the current development only. It allows quickly commenting/uncommenting logs

File: lib/supavisor/handlers/proxy/client.ex [hunk 13]
```diff
@@ -0,0 +1,673 @@
...
+        :internal,
+        {:client, {:hello, {type, {user, tenant_or_alias, db_name}}}},
+        :exchange,
+        %{sock: sock} = data
+      ) do
+    sni_hostname = HH.try_get_sni(sock)
+
+    case Tenants.get_user_cache(type, user, tenant_or_alias, sni_hostname) do
+      {:ok, info} ->
+        db_name = if(db_name != nil, do: db_name, else: info.tenant.db_database)
```
[the remark this answers is not in the document]
Comment (abc3) [reply lib/supavisor/handlers/proxy/client.ex:141]: I feel it makes readability harder

File: lib/supavisor/handlers/proxy/client.ex (hunk 3 above)
Comment (abc3) [reply lib/supavisor/handlers/proxy/client.ex:655]: yep, forgot to change

File: lib/supavisor/handlers/proxy/db.ex (hunk 4 above)
Comment (abc3) [reply lib/supavisor/handlers/proxy/db.ex:156]: can you please clarify?

File: lib/supavisor/handlers/proxy/handler.ex (hunk 6 above)
Comment (abc3) [reply lib/supavisor/handlers/proxy/handler.ex:41]: 100%

**Comment by abc3** [comment]
> Should we consider using defdelegate in Supavisor.Handlers.Proxy.Handler?

@hauleth wdyt?

File: lib/supavisor/helpers.ex (hunk 9 above)
Comment (hauleth) [reply lib/supavisor/helpers.ex:364]: No need for dynamic lookup going through `String.Chars`.

Comment (abc3) [reply lib/supavisor/helpers.ex:364]: Yeah, but it is called once before start and doesn't influence performance

File: lib/supavisor/handlers/proxy/db.ex (hunk 4 above)
Comment (hauleth) [reply lib/supavisor/handlers/proxy/db.ex:156]: That function has 140 lines of code.

File: lib/supavisor/handlers/proxy/client.ex (hunk 13 above)
[the remark this answers is not in the document]
Comment (hauleth) [reply lib/supavisor/handlers/proxy/client.ex:141]: And `!!` operator does not?

File: lib/supavisor/helpers.ex (hunk 9 above)
Comment (abc3) [reply lib/supavisor/helpers.ex:364]: okay, need to replace 
```
Name                     ips        average  deviation         median         99th %
list to string       12.20 M       81.97 ns ±25614.07%          42 ns          84 ns
# { }                 9.82 M      101.87 ns ±30388.30%          83 ns          84 ns

Comparison:
list to string       12.20 M
# { }                 9.82 M - 1.24x slower +19.89 ns
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
[os_mon] memory supervisor port (memsup): Erlang has closed
```

File: lib/supavisor/handlers/proxy/db.ex (hunk 4 above)
Comment (abc3) [reply lib/supavisor/handlers/proxy/db.ex:156]: ahh, gotcha

File: lib/supavisor/handlers/proxy/client.ex (hunk 13 above)
[the remark this answers is not in the document]
Comment (abc3) [reply lib/supavisor/handlers/proxy/client.ex:141]: :sweat_smile:

File: lib/supavisor/handlers/proxy/client.ex (hunk 10 above)
[the remark this answers is not in the document]
Comment (hauleth) [reply lib/supavisor/handlers/proxy/client.ex:17]: @abc3 it makes it easier for my eyes to parse. It also make it easier to grep through codebase when it is needed.

**Comment by hauleth** [comment]
About `defdelegate`? No strong opinion on that matter.

File: lib/supavisor/handlers/proxy/client.ex [hunk 14]
```diff
@@ -0,0 +1,662 @@
...
+
+            Telem.client_join(:fail, id)
+            {:stop, {:shutdown, :address_not_allowed}}
+
+          true ->
+            new_data = update_user_data(data, info, user, id, db_name, mode)
+
+            key = {:secrets, tenant_or_alias, user}
+
+            case auth_secrets(info, user, key, :timer.hours(24)) do
```
Comment (hauleth) [inline lib/supavisor/handlers/proxy/client.ex:175]: Another deeply nested conditional.

**Review by hauleth — CHANGES_REQUESTED** [review]
It would be handy to split some stuff to separate PRs.

File: mix.exs [hunk 15]
```diff
@@ -60,7 +60,6 @@ defmodule Supavisor.MixProject do
       # TODO: point it to Supabase fork of prom_ex when available
       {:prom_ex, github: "hauleth/prom_ex", branch: "ft/add-peep-storage"},
       {:open_api_spex, "~> 3.16"},
-      {:burrito, github: "burrito-elixir/burrito"},
```
[the remark this answers is not in the document]
Comment (abc3) [reply mix.exs:63]: it already merged in v2 as a separate pr https://github.com/supabase/supavisor/pull/395

File: typos.toml
[the remark this answers is not in the document]
Comment (abc3) [reply typos.toml:1]: very odd

**Comment by abc3** [comment]
#401

**Pull request closed without merging**